p2panda-rs/src/atomic/message.rs

```diff
@@ -0,0 +1,379 @@
+use anyhow::bail;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use serde_repr::{Deserialize_repr, Serialize_repr};
+use std::collections::HashMap;
+use thiserror::Error;
+
+use crate::atomic::{Hash, MessageEncoded, Validation};
+use crate::Result;
+
+/// Message format versions to introduce API changes in the future.
+#[derive(Clone, Debug, PartialEq, Serialize_repr, Deserialize_repr)]
+#[serde(untagged)]
+#[repr(u8)]
+pub enum MessageVersion {
+    /// All messages are currently implemented against this first version.
+    Default = 1,
+}
+
+impl Copy for MessageVersion {}
+
+/// Messages are categorized by their `action` type.
+///
+/// An action defines the message format and if this message creates, updates or deletes a data
+/// instance.
+#[derive(Clone, Debug, PartialEq)]
+pub enum MessageAction {
+    /// Message creates a new data instance.
+    Create,
+
+    /// Message updates an existing data instance.
+    Update,
+
+    /// Message deletes an existing data instance.
+    Delete,
+}
+
+impl Serialize for MessageAction {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(match *self {
+            MessageAction::Create => "create",
+            MessageAction::Update => "update",
+            MessageAction::Delete => "delete",
+        })
+    }
+}
+
+impl<'de> Deserialize<'de> for MessageAction {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let s = String::deserialize(deserializer)?;
+
+        match s.as_str() {
+            "create" => Ok(MessageAction::Create),
+            "update" => Ok(MessageAction::Update),
+            "delete" => Ok(MessageAction::Delete),
+            _ => Err(serde::de::Error::custom("unknown message action")),
+        }
+    }
+}
+
+impl Copy for MessageAction {}
+
+/// Enum of possible data types which can be added to the messages fields as values.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum MessageValue {
+    /// Basic `boolean` value.
+    Boolean(bool),
+
+    /// Basic signed `float` value.
+    Float(f64),
+
+    /// Basic signed `integer` value.
+    Integer(i64),
+
+    /// Basic `string` value.
+    Text(String),
+}
```

Comment:
There's no reference type! :o

Reply:
How does it look like?

Reply:
`Relation(Hash)`? Does that make sense?

Reply:
> Does that make sense?

Maybe its a tuple of `Relation(Hash, Schema)`?

Reply:
The schema of the relation target is defined by the schema of the message this `MessageValue` would be part of. What do you have in mind when you say legacy schemas?

Reply:
I see, then we only need `Hash` :+1:

Reply:
Maybe instead of `Hash` we should also add an `InstanceID` type that wraps a hash but offers more functionality related to the instance it's referring to. I would prefer not to also do that in this PR though.

Reply:
Ja that would make sense, better than wrapping wrapped wrappers might be to introduce a `Hash` Trait which gets implemented by `Schema`, `InstanceId` and `EntryHash` or similar?

Reply:
Would that be *the* Hash trait or our own, which would also ensure a YamfHash etc?

Reply:
Our own with `encode`, `decode` methods etc.

Reply:
Better name needed probably.

Reply:
We can continue discussing that [in this issue](https://github.com/p2panda/p2panda/issues/32) eventually.